File: cpp/units.cpp
#include "time_nanos.h"
#include "time_stamp.h"

#include <vector>

using namespace std;


void test_blog_examples()
{
	cout << "Showing blog examples." << endl;

	auto a = "2012.01.01"_date;
	cout << "Example 1: " << a << endl;
	cout << "Example 2:" << endl;
	cout << "\t" << a + 1_d << endl;
	cout << "\t" << "2012.01.01"_date + 1_d << endl;
	cout << "Example 3: " <<
		"2012.01.01"_date + 1_m + 1_d + 1_h + 15_i + 17_s << endl;

	auto r = range(0, 10);
	auto offsets = map(r, [] (int i) {
		return (1_m + 1_d + 1_h + 15_i + 17_s) * i;
		});
	auto values = map(offsets, [] (Nano x) { return "2012.01.01"_date + x; });
	cout << "Example 4: " << values << endl;

	cout << "…using helper procs… "
		<< "2012.01.01"_date + (1_m + 1_d + 1_h + 15_i + 17_s) * range(0, 10)
		<< endl;

	cout << "Example 5 b[week]: " <<
		map(values, [] (Stamp x) { return x.week(); }) << endl;
	cout << "Example 5 b[second]: " <<
		map(values, [] (Stamp x) { return x.second(); }) << endl;

	cout << endl << "Did all examples." << endl;
}

int main()
{
	test_nanos();
	test_stamp();
	test_blog_examples();
}

File: nim/units.nim
# Implementation of seconds, used for time differences.
import time_nanos, time_stamp, sequtils

proc bonusWeek(x: Stamp): string {.procvar.} = "^" & $week(x) & "^"

proc test_blog_examples() =
  echo "Showing blog examples.\n"

  let a = d"2012.01.01"
  echo "Example 1: ", a
  echo "Example 2:"
  echo "\t", a + 1.d
  echo "\t", "2012.01.01".date + 1.d
  echo "Example 3: ", "2012.01.01".date + 1.m + 1.d + 1.h + 15.i + 17.s

  let
    r = to_seq(0 .. <10)
    offsets = r.map_it(Nano, (1.m + 1.d + 1.h + 15.i + 17.s) * it)
    values = offsets.map_it(Stamp, d"2012.01.01" + it)

  echo "Example 4: ", values

  echo "…again but compressed… ", to_seq(0 .. <10)
      .map_it(Nano, (1.m + 1.d + 1.h + 15.i + 17.s) * it)
      .map_it(Stamp, "2012.01.01".date + it)

  echo "…using helper procs… ",
    d"2012.01.01" + (1.m + 1.d + 1.h + 15.i + 17.s) * (0 .. <10)

  echo "Example 5 b[week]: ", values[week]
  echo "Example 5 b[second]: ", values[second]

  echo "Extra points b[bonusWeek]: ", values[bonusWeek]

  echo "\nDid all examples."

proc main() =
  test_seconds()
  test_stamps()
  test_blog_examples()

when isMainModule: main()
